core/lib/zksync_core/src/state_keeper/batch_executor/tests/read_storage_factory.rs

```diff
@@ -0,0 +1,104 @@
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::{runtime::Handle, sync::watch};
+use zksync_dal::ConnectionPool;
+use zksync_state::{open_state_keeper_rocksdb, PostgresStorage, RocksdbStorageBuilder};
+
+use crate::state_keeper::state_keeper_storage::{PgOrRocksdbStorage, ReadStorageFactory};
+
+#[derive(Debug, Clone)]
+pub struct PostgresFactory {
+    pool: ConnectionPool,
+}
+
+#[async_trait]
+impl ReadStorageFactory for PostgresFactory {
+    async fn access_storage<'a>(
+        &'a self,
+        rt_handle: Handle,
+        _stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        let mut connection = self
+            .pool
+            .access_storage()
+            .await
+            .context("Failed accessing Postgres storage")?;
+
+        let snapshot_recovery = connection
```

Comment:
This looks incorrect; looks like you've fixed snapshot-related logic in the main code, but not here. Wouldn't it be possible to reuse the code from `AsyncRocksdbCache` here and in `impl ReadStorageFactory for RocksdbFactory` so that such mistakes are eliminated by design?